zip-0317.rst

```diff
@@ -262,75 +262,81 @@ following section is planned to be implemented by `zcashd` and `zebrad`.
 Recommended algorithm for block template construction
 '''''''''''''''''''''''''''''''''''''''''''''''''''''
 
-Define a constant :math:`weight\_cap = 4`.
+Define constants :math:`weight\_ratio\_cap = 4` and
+:math:`block\_unpaid\_action\_limit = 50\!`.
 
 Let :math:`conventional\_fee(tx)` be the conventional fee for transaction
 :math:`tx` calculated according to the section `Fee calculation`_.
 
+Let :math:`unpaid\_actions(tx) = \mathsf{max}\!\left(0,\, \mathsf{max}(grace\_actions,\, tx.\!logical\_actions) - \mathsf{floor}\!\left(\frac{tx.fee}{marginal\_fee}\right)\right)\!`.
+
+Let :math:`block\_unpaid\_actions(block) = \sum_{tx \,\in\, block}\, unpaid\_actions(tx)`.
+
 The following algorithm is RECOMMENDED for constructing block templates
 from a set of transactions in a node's mempool:
 
 1. For each transaction :math:`tx` in the mempool, calculate
-   :math:`tx.\!weight = \mathsf{min}\!\left(\frac{tx.fee}{conventional\_fee(tx)}, weight\_cap\right)\!`.
+   :math:`tx.\!weight\_ratio = \mathsf{min}\!\left(\frac{\mathsf{max}(1,\, tx.fee)}{conventional\_fee(tx)},\, weight\_ratio\_cap\right)\!`
+   and add the transaction to the set of candidate transactions.
 
-2. Repeat while there is any mempool transaction that pays at least the
-   conventional fee, is within the block sigop limit [#sigop-limit]_, and
-   fits in the block:
+2. Repeat while there is any candidate transaction that pays at least the
+   conventional fee:
 
    a. Pick one of those transactions at random with probability in direct
-      proportion to its weight, and add it to the block.
-
-3. Let :math:`N` be the number of remaining transactions with :math:`tx.\!weight < 1\!`.
-   Calculate their sum of weights, call this :math:`remaining\_weight\!`.
+      proportion to its weight ratio, and remove it from the set of candidate
```

Comment:
"In proportion to its weight ratio" -> I'd suggest "In proportion to $tx.capped_weight_ratio$" as referencing a specific variable is more explicit than describing that variable in prose. (Same for Step 3.)